README.md
![SRF.ch build status](https://github.com/kaespi/srfchwx/actions/workflows/srfchwx_ci.yml/badge.svg) ![GitHub](https://img.shields.io/github/license/kaespi/srfchwx) ![GitHub tag (latest by date)](https://img.shields.io/github/v/tag/kaespi/srfchwx)


# SRF.ch Firefox extension

This is a Firefox extension to obtain the URLs of video- and audio-files on the Swiss national radio and television webpage (www.srf.ch), using the new WebExtensions format.

The extension is published in the official Firefox Add-Ons store, but in the non-checked part: https://addons.mozilla.org/en-US/firefox/addon/srfch/.

## How it's done

This a short description of how the media URLs are derived.

### Step 1: get the media's unique ID

Each video/audio has its own unique ID. The ID follows the regex `[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}`, e.g. 9f755253-7ccd-40f9-81a9-15f6b73dd4c7. The task of the browser extension is to find this ID in the page's DOM tree starting from the element which was right-clicked. Naturally, the page's look changes every now and then, hence there are various ways to obtain the ID and they need to be updated sometimes.

### Step 2: get the media's different streams

Once the ID was obtained, the different streams for the video/audio can be found in a JSON file from srgssr.ch server. Namely the JSON file can be found
 * for srf.ch videos: `https://il.srgssr.ch/integrationlayer/2.0/mediaComposition/byUrn/urn:srf:video:<ID>.json`;
 * for srf.ch audios: `https://il.srgssr.ch/integrationlayer/2.0/srf/mediaComposition/audio/<ID>.json`
 * for rsi.ch stuff: `http://il.srgssr.ch/integrationlayer/1.0/ue/rsi/video/play/<ID>.json`

Replace the `<ID>` above by the media's unique ID as obtained in step 1.

The JSON file contains a lot of meta information about the media, such as the title, the time it was broadcast, a short description etc., see an example below
<details>
    <summary>show example JSON file excerpts</summary>

```
{
  "Video": {
    "id": "9f755253-7ccd-40f9-81a9-15f6b73dd4c7",
    "modifiedDate": "2021-11-29T08:51:14+01:00",
    "urn": "urn:srf:ais:video:9f755253-7ccd-40f9-81a9-15f6b73dd4c7",
    "displayable": true,
    "createdDate": "2021-11-26T11:17:06+01:00",
    "assetSetId": "9ac55409-7209-43de-8d9c-03cea6b2ed42",
    "position": 0,
    "assetSubSetId": "EPISODE",
    "validFrom": "2021-11-28T22:20:00+01:00",
    "validTo": "2022-05-27T23:59:00+02:00",
    "noEmbed": true,
    "AssetMetadatas": {
      "AssetMetadata": [
        {
          "id": "aea188a1-2b5b-4a55-92db-b90a45c91c5a",
          "modifiedDate": "2021-11-29T08:51:14+01:00",
          "createdDate": "2021-11-26T11:17:06+01:00",
          "description": "Auf dem Simplon kommt es zu einer [...]",
          "lead": "Eine Konfrontation auf dem [...]",
          "title": "Tomatsossu",
          "assetId": "9f755253-7ccd-40f9-81a9-15f6b73dd4c7",
          "usage": "DEFAULT"
        }
      ]
    },
    ...
    ...
    "Playlists": {
      "@availability": "ONDEMAND",
      "Playlist": [
       ...
       ...
        {
          "@segmentation": "LOGICAL",
          "@protocol": "HTTP-HLS",
          "url": [
            {
              "@quality": "SD",
              "text": "https://srfvodhd-vh.akamaihd.net/i/vod/tschugger/2021/11/tschugger_20211125_135500_7372782_v_webcast_h264_,q40,q10,q20,q30,q50,.mp4.csmil/master.m3u8"
            },
            {
              "@quality": "HD",
              "text": "https://srfvodhd-vh.akamaihd.net/i/vod/tschugger/2021/11/tschugger_20211125_135500_7372782_v_webcast_h264_,q40,q10,q20,q30,q50,q60,.mp4.csmil/master.m3u8"
            }
          ]
        }
      ]
...
...
```
</details>

In this JSON file we can find the playlist file for the media, the `master.m3u8`. This leads us to the third step.

### Step 3: extract the different streams from master.m3u8

The playlist file (`master.m3u8`) is then downloaded and parsed. It's a plain ASCII file. It contains addresses of the streams in various qualities. See the example below:

```
#EXTM3U
#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=1327000,RESOLUTION=640x360,CODECS="avc1.77.30, mp4a.40.2",CLOSED-CAPTIONS=NONE
https://srfvodhd-vh.akamaihd.net/i/vod/tschugger/2021/11/tschugger_20211125_135500_7372782_v_webcast_h264_,q40,q10,q20,q30,q50,q60,.mp4.csmil/index_0_av.m3u8?null=0
#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=148000,RESOLUTION=320x180,CODECS="avc1.66.30, mp4a.40.2",CLOSED-CAPTIONS=NONE
https://srfvodhd-vh.akamaihd.net/i/vod/tschugger/2021/11/tschugger_20211125_135500_7372782_v_webcast_h264_,q40,q10,q20,q30,q50,q60,.mp4.csmil/index_1_av.m3u8?null=0
#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=348000,RESOLUTION=480x272,CODECS="avc1.66.30, mp4a.40.2",CLOSED-CAPTIONS=NONE
https://srfvodhd-vh.akamaihd.net/i/vod/tschugger/2021/11/tschugger_20211125_135500_7372782_v_webcast_h264_,q40,q10,q20,q30,q50,q60,.mp4.csmil/index_2_av.m3u8?null=0
#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=695000,RESOLUTION=512x288,CODECS="avc1.77.30, mp4a.40.2",CLOSED-CAPTIONS=NONE
https://srfvodhd-vh.akamaihd.net/i/vod/tschugger/2021/11/tschugger_20211125_135500_7372782_v_webcast_h264_,q40,q10,q20,q30,q50,q60,.mp4.csmil/index_3_av.m3u8?null=0
#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=2127000,RESOLUTION=960x544,CODECS="avc1.77.30, mp4a.40.2",CLOSED-CAPTIONS=NONE
https://srfvodhd-vh.akamaihd.net/i/vod/tschugger/2021/11/tschugger_20211125_135500_7372782_v_webcast_h264_,q40,q10,q20,q30,q50,q60,.mp4.csmil/index_4_av.m3u8?null=0
#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=3627000,RESOLUTION=1280x720,CODECS="avc1.77.30, mp4a.40.2",CLOSED-CAPTIONS=NONE
https://srfvodhd-vh.akamaihd.net/i/vod/tschugger/2021/11/tschugger_20211125_135500_7372782_v_webcast_h264_,q40,q10,q20,q30,q50,q60,.mp4.csmil/index_5_av.m3u8?null=0
#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=48000,CODECS="mp4a.40.2",CLOSED-CAPTIONS=NONE
https://srfvodhd-vh.akamaihd.net/i/vod/tschugger/2021/11/tschugger_20211125_135500_7372782_v_webcast_h264_,q40,q10,q20,q30,q50,q60,.mp4.csmil/index_1_a.m3u8?null=0

```

In this last step this file is parsed, the meta information (such as bandwidth, resolution etc.) is available from the lines starting with `#`. This is then presented to the user which can then find for different qualities the actual URLs to play/download the video or audio stream.

## How it works in Firefox

This extension simply extracts the URLs of the video- and audio-files on the SRF (Schweizer Radio und Fernsehen, Swiss german national television and radio) and the RSI (Radiotelevisione Svizzera, Swiss italian national television and radio) webpage, www.srf.ch and www.rsi.ch respectively. After right-clicking on a video or audio link (or a banner which this extension sometimes adds to the pages) you can start extracting/grabbing the corresponding URLs by clicking on the context menu entry ("Extract SRF/RSI URLs"). If some URLs are then found an SRF/RSI icon will popup in the address bar of the browser. This one adds a dropdown menu from where you can copy the URLs by clicking on any of the dropdown entries.

Some of the content on the webpage is geo-blocked. This means that  the URLs can't be extracted. After some timeout the icon in the addressbar will appear as grayed-out icon. Other content is time blocked, meaning that the media file is only available over a fixed time-window. This is particularly the case for Tatort, Spielfilm and DOK videos.

Starting with version 1.1.5 the add-on extracts the URLs of the M3U playlists. These entries show up with their M3U description in the context menu rather than the RTMP URL like the other entries. They can be helpful because M3U playlists are often available even if the stream is geo- or time-blocked. Instructions on how to download these files is given further below.

## How to use the extracted URLs

The media files can (only?) be easily downloaded using open source tools. Media files for which M3U playlists are available can be directly played with VLC or any other player understanding the M3U format. Others, like those with RTMP probably need special software such as FLVstreamer.

To download any streams/files ffmpeg is very useful tool, using the following command (kudos to [flip](https://oinkzwurgl.org/))

    ffmpeg -i <url> -codec copy <outputfile>


The downloaded files can be played with
 * Video:
   most (if not all!?) video files are of the format H264 with MPEG AAC audio. They can be viewed with the VLC player.
 * Audio:
   the audio files are of MPEG audio format (not mp3!). They can also be played with VLC.

Helpful software:
 * [VLC player](http://www.videolan.org/vlc)
 * [FLVstreamer](http://www.nongnu.org/flvstreamer)
 * [FFmpeg](http://www.ffmpeg.org/)


Thanks to a hint from a user, the following command could be used to download an M3U file on Linux:

    wget -qO - `wget -qO - <url>`

Or - as said above - using ffmpeg:

    ffmpeg -i <url> -codec copy <outputfile>

-Kaspar Giger <sftv@kgmw.ch>


changelog.txt
Changelog SRF.ch
================

srfch 2.0.4
    * Adapt URL for JSON file to extract media meta data on srf.ch webpage
      (https://il.srgssr.ch/integrationlayer/2.0/mediaComposition/byUrn/urn:srf:video:<ID>.json
      instead of previous http://il.srgssr.ch/integrationlayer/1.0/ue/srf/video/play/<ID>.json)

srfch 2.0.3
    * Adapt media extraction to the latest appearance of the srf.ch webpage

srfch 2.0.2
    * Fix appearance of banners on srf.ch/play and rsi.ch pages

srfch 2.0.1
    * Only refer to this add-ons internal files in the banners on rsi.ch and srf.ch/play
    * Better triggering adding the banner on srf.ch/play and rsi.ch pages

srfch 2.0.0
    * Ported the add-on to the new WebExtensions standard (which resulted in some small
      changes in its appearance and behavior)

srfch 1.2.3
    * Made add-on work with https:// URLs (in a trivial way)

srfch 1.2.2
    * Fixed bug in srf.ch URL retrieval

srfch 1.2.1
    * Added support for RSI (www.rsi.ch), thanks to Rossano Gobbi
    * Removed deprecated call to newChannelFromURI and changed to newChannelFromURI2

srfch 1.1.16
    * SRF changed the URL from where to get information about a media using
      the unique ID

srfch 1.1.15
    * better get the media file's unique ID from the srf.ch webpage by following
      "hidden" links

srfch 1.1.14
    * also grab the media file's unique ID from the data-urn attribute (makes
      URL extraction work again on some Javascript click style links)

srfch 1.1.13
    * only implementation improvements

srfch 1.1.12
    * Added a link to the srf.ch/play... pages to keep compatibility with
      the new srf.ch player webpage (since around July 2015)

srfch 1.1.11
    * Added compatibility for new srf.ch URLs (since end of January 2015)
    * Added guessing of medium high qualitz M3U media stream files (q50)

srfch 1.1.10

    * Added guessing of high quality M3U media stream files, giving access
      to HQ videos

srfch 1.1.9

    * Bugfix: Audio URLs properly read out

srfch 1.1.8

    * Bugfix: Removed debug outputs

srfch 1.1.7

    * Bugfix: M3U entries are no more overwritten by "regular" RTMP URLs, i.e.
      all M3U entries are always output
    * Prevention of displaying duplicate M3U entries
    * Added 5 new potential entries

srfch 1.1.6

    * Bugfix: M3U entries are no more overwritten by guessed RTMP URLs, i.e.
      all M3U entries are always output

srfch 1.1.5

    * Adds M3U entries to the context menu for each media (due to RTMP
      sometimes failing - as of September 2013)

srfch 1.1.4

    * For media files which have no RTMP URL available the add-on tries
      to find the URLs of M3U playlists which can be played with any
      media player capable of handling M3U playlists (such as e.g. VLC).
      Those M3U playlist URLs show up with leading "M3U:" and the M3U
      description of the media file (look for the BANDWIDTH entry!)

srfch 1.1.3

    * The URLs of HQ-videos in the archive are guessed (because usually they
      are available from the RTMP-server, although not indicated)
    * URLs which are guessed are marked with a leading asterisk * in the
      context menu

srfch 1.1.2

    * Now also the extraction of URLs of videos in the archive (old
      sf.tv videoplayer) available
    * Added the capability to extract URLs of spielfilm videos
      (feature films) which are no longer available directly from the web

srfch 1.1.1

    * Added the capability to extract URLs of DOK videos (documentary
      movies) which are no longer available directly from the web

srfch 1.1.0

    * Added the option to extract URLs of Tatort videos that are
      time-blocked from 06:00 - 20:00 (CET)

srfch 1.0.1

    * Removed unnecessary content
    * Moved all content and locale to the chrome directory

srfch 1.0.0

    * Initial release.

src/popup/srfch_copy_link.js
// SRF.ch Firefox extension, popup script
// Copyright (C) 2017 Kaspar Giger (sftv@kgmw.ch)
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software Foundation,
// Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA


// id of the tab which was opened
var popupTabId = null;

// listens for the mouse click on an entry of the popup
document.addEventListener("click", function(e) {
    if (e.target.classList.contains("media-select"))
    {
        var url = e.target.dataset.url;
        if (url)
        {
            // we need to select the text somehow. To do so we quickly add a textarea element
            // to the DOM tree, fill it with the URL, select its text and copy this one:
            //  1. create the textarea element
            var textarea = window.document.createElement("textarea");
            textarea.className = "url-txt"; // class 'url-txt' will make it of height 1px (such that it won't be noticeable)
            var txtUrl = window.document.createTextNode(url);
            //  2. fill the textarea with the URL
            textarea.appendChild(txtUrl);
            window.document.body.appendChild(textarea);

            //  3. select the textarea's contents (i.e. the URL)
            textarea.select();
            //  4. copy the selected text to clipboard
            document.execCommand("Copy");

            //  5. delete the textarea again
            textarea.parentNode.removeChild(textarea);

            window.close();
            return;
        }
    }
});

/*
    onGotBg(page):
    Function called after the popup is opened and a background page is available. This
    function basically gets the URLs the background page has extracted and shapes the
    entries of the popup
*/
function onGotBg(page)
{
    // only add some elements to the popup if URLs are available...
    if (page.media[popupTabId].length > 0)
    {
        var title = page.mediaTitle[popupTabId];
        if (!title)
        {
            title = browser.i18n.getMessage("pageActionUntitled");
        }
        var divTitle = window.document.createElement("div");
        divTitle.className = "media-title";
        var txtTitle = window.document.createTextNode(title + ": (" + browser.i18n.getMessage("pageActionCopyInstr") + ")");
        divTitle.appendChild(txtTitle);
        window.document.body.appendChild(divTitle);

        for (var k=0; k<page.media[popupTabId].length; k++)
        {
            var div = window.document.createElement("div");
            div.className = "media-select";
            div.dataset.url = page.media[popupTabId][k].url;
            var desc = page.media[popupTabId][k].desc;
            // prevent oversized descriptions...
            if (desc.length > 90)
            {
                desc = desc.substr(0,70) + "....." + desc.substr(-19);
            }
            var txt = window.document.createTextNode(desc);
            div.appendChild(txt);
            window.document.body.appendChild(div);
        }
    }
}

/*
    onBgError(page):
    Function called after the popup is opened and the background page is not available.
*/
function onBgError(error)
{
    console.log("[srfch] error getting bg page");
}

// the code from here on is executed when the popup is opened (i.e. the user clicks the icon)

// clear the existing nodes: title...
var divs = window.document.getElementsByClassName("media-title");
while (divs[0])
{
    divs[0].parentNode.removeChild(divs[0]);
}

// ... and selectors
var divs = window.document.getElementsByClassName("media-select");
while (divs[0])
{
    divs[0].parentNode.removeChild(divs[0]);
}

// query the current active tab ID. Without knowing it we cannot continue
// here because the background page stores information for each tab individually
var gettingActiveTab = browser.tabs.query({active: true, currentWindow: true});
gettingActiveTab.then((tabs) => {
    popupTabId = tabs[0].id;

    var getting = browser.runtime.getBackgroundPage();
    getting.then(onGotBg, onBgError);
});


src/js/srfch_content.js
// SRF.ch Firefox extension, content script
// Copyright (C) 2017 Kaspar Giger (sftv@kgmw.ch)
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software Foundation,
// Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA


// upon loading the content script we send an empty message to the background
// script such that the background script's data for this tab is initialized
// and the pageAction safely hidden
var contentLoaded = 0;

// initialize the data for this tab (and remove the pageAction popup initially)
if (!contentLoaded)
{
    // only send this message once (otherwise already existing data could be
    // reset to zero)
    contentLoaded = 1;
    browser.runtime.sendMessage({});
}

document.addEventListener("load", function(event) {
    // on srf.ch/play pages the video usually just starts without given this add-on the
    // possibility to extract the video's URL, although the id of the video is nicely
    // contained in the URL. If this is the case then add a banner on top of the video
    // to make this add-on work even in this situation
    if (document.documentURI
        && document.documentURI.match(/srf\.ch\/play\/.*\/video\/.*id=[-a-zA-Z0-9]{36}/g))
    {
        addSrfPlayBanner();
    }
    // the rsi.ch page is a bit a tricky case, because it embeds the videos in iframes
    // whose source is on srgssr.ch. Therefore we have some cross-domain problem. Since
    // the content script is not allowed to interfere with the iframes and cannot catch
    // mouse events there, we just have to add some visual element close to the iframe
    // (similar like for the srf.ch/play page)
    else if (document.documentURI && (document.documentURI.indexOf("rsi.ch") >= 0))
    {
        var iframes = document.getElementsByTagName("iframe");
        if (iframes)
        {
            for (var k=0; k<iframes.length; k++)
            {
                if (iframes[k].src.indexOf("srgssr.ch") >= 0)
                {
                    var urnrsi = iframes[k].src.match(/urn:rsi:video:[0-9]{7}/);
                    if (urnrsi.length==1)
                    {
                        addRsiVideoBanner(iframes[k], urnrsi[0]);
                    }
                }
            }
        }
    }

    // send an empty message to the background script. This makes the toolbar-popup
    // being disabled (by default) and the context menu being removed (if it exists)
    if (!contentLoaded)
    {
        // only send this message once (otherwise already existing data could be
        // reset to zero)
        contentLoaded = 1;
        browser.runtime.sendMessage({});
    }
}, true);

window.addEventListener("beforeunload", function(event) {
    // send an empty message to the background script. This makes the toolbar-popup
    // being disabled (by default) and the context menu being removed (if it exists)
    browser.runtime.sendMessage({});
}, true);

window.addEventListener("mousedown", function(event) {
    // right click
    // (this will probably trigger the context menu to pop up)
    if (event.button == 2)
    {
        var idStr = "";

        // check if (right-) clicked on a link
        if (event.target && event.target.closest("a"))
        {
            // check if the "urn" data attribute is set, then we can easily extract the
            // media's ID from it
            var elemA = event.target.closest("a");
            if (elemA.dataset.urn)
            {
                idStr = elemA.dataset.urn;
            }
            else
            {
                // sometimes the ID is hidden somehow behind another node in the
                // DOM tree which is linked via the "href"-attribute. Check this option...
                var hrefAttr = elemA.getAttribute("href");
                if (hrefAttr && hrefAttr.substring(0,1)==="#")
                {
                    var promoId = hrefAttr.substring(1);

                    // check if a node with this id exists
                    var promoNode = document.getElementById(promoId);
                    if (promoNode)
                    {
                        // check if any of the child elements there's the 'data-urn'
                        // attribute set
                        var promoChildren = promoNode.getElementsByTagName("*");

                        for (var i=0; i<promoChildren.length; i++)
                        {
                            if(promoChildren[i].dataset && promoChildren[i].dataset.urn)
                            {
                                idStr = promoChildren[i].dataset.urn;
                                break;
                            }
                        }
                    }
                }

                // if we didn't find the "urn" data attribute yet, then we can try to extract
                // the ID from the "href" attribute of the link node. To do so, try to find
                // the ID using a regex
                if (idStr=="" && hrefAttr.match(/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/))
                {
                    idStr = hrefAttr;
                }
            }
        }
        else
        {
            // looks like it's the "new" way how this is represented on the srf.ch webpage (found it first in
            // September 2021). We might find the ID in one of the target's parent which has it encoded in the
            // "id" attribute or the .data.assetid entry.
            idStr = getIdWithIdAttr(event.target.parentNode);

            if (!idStr)
            {
                // ok, we didn't find the ID in one of the target's parents. We might be on a srf.ch/play/...
                // page and the user clicked the "main video element". If that's the case, let's apply some magic to
                // get the ID. It might be encoded in some element which is used to handle the player controls
                idStr = getIdWithPlayerCtrl(event.target);
            }
        }

        // if the string with the ID was set then we can forward this information (via a
        // message) to the background script which can then do the actual extraction of
        // the URLs
        browser.runtime.sendMessage({srfchId: idStr});
    }
}, true);

/*
    getIdWithIdAttr():
    Try to extract the ID from from any of the parents of "startingNode" by looking for an
    "id"-attribute.
*/
function getIdWithIdAttr(startingNode)
{
    // Let's try to find the ID in one of the node's parent which has it encoded in the 
    // "id" attribute or the .data.assetid entry.
    var idStr = "";
    var testNode = startingNode;
    for (let i=0; i<6; i++)
    {
        if (testNode.hasAttribute("id"))
        {
            var idAttr = testNode.getAttribute("id");
            var idMatch = idAttr.match(/player-wrapper__urn-srf-([a-z]+)-([0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})/);
            if (idMatch)
            {
                idStr = "urn:srf:" + idMatch[1] + ":" + idMatch[2];
                break;
            }
        }

        if (testNode.dataset.assetid &&
            testNode.dataset.assetid.match(/urn:srf:[a-z]+:[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/))
        {
            idStr = testNode.dataset.assetid;
            break;
        }

        if (testNode.parentNode)
        {
            // if it's not this element, it could be its parent where the information is available
            testNode = testNode.parentNode;
        }
        else
        {
            // no more climbing up the DOM tree
            break;
        }
    }

    return idStr;
}

/*
    getIdWithPlayerCtrl():
    Try to obtain the media ID by crawling up and down the DOM tree starting from some
    note "startingNode". The ID might be encoded in an <a ...> node's href attribute.
*/
function getIdWithPlayerCtrl(startingNode)
{
    var idStr = "";
    var parentNode = startingNode;
    for (let levelUp=0; levelUp<6 && !idStr; levelUp++)
    {
        idStr = getIdStrFromAHref(parentNode, 4, 0);

        if (parentNode.parentNode)
        {
            // if it's not this element, it could be its parent where the information is available
            parentNode = parentNode.parentNode;
        }
        else
        {
            // no more climbing up the DOM tree
            break;
        }
    }

    return idStr;
}

/*
    getIdStrFromAHref():
    Tries to extract the media ID from this node (if it's an <a ...> node and has a href
    attribute with the ID encoded). Otherwise recursively processes all the node's children.
*/
function getIdStrFromAHref(node, maxLevelDown, levelDown)
{
    var idStr = "";
    if (node.nodeName.toLowerCase()=="a")
    {
        // found an <a ...> node. Let's see if it contains in its href-attribute
        // the potential ID
        var hrefAttr = node.getAttribute("href");
        var idMatch = hrefAttr.match(/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/);
        if (idMatch)
        {
            idStr = "urn:srf:video:" + idMatch[0];
        }
    }
    else
    {
        if (node.hasChildNodes())
        {
            levelDown++;
            let children = node.childNodes;
            for (let k=0; k < children.length && !idStr; k++)
            {
                idStr = getIdStrFromAHref(children[k], maxLevelDown, levelDown);
            }
        }
    }

    return idStr;
}

/*
    addRsiVideoBanner():
    Since on rsi.ch the videos are loaded in iframes from a different domain (not rsi.ch), we cannot
    handle the iframe properly. Therefore we need to add a banner to the webpage, right in front of
    the iframe...
*/
function addRsiVideoBanner(iframe, urn)
{
    // we should not add the same banner multiple times. To prevent so we add the div
    // an id with the number in the urn
    var bannerId = "";
    var urnDigitMatches = urn.match(/\d+/g);
    if (urnDigitMatches)
    {
        bannerId = "srfchaddonbanner";
        for (var k=0; k<urnDigitMatches.length; k++)
        {
            bannerId += urnDigitMatches[k];
        }
    }
    if (bannerId && document.getElementById(bannerId))
    {
        return;
    }

    var divIframe = iframe.parentNode;
    var nodeParent = divIframe.parentNode;

    // add the following HTML code to the current document's DOM tree:
    // <div style="text-align:center">
    //     <img class="srfchaddon" src="http://addons.cdn.mozilla.net/user-media/addon_icons/413/413748-64.png" />
    //       Download URLs for this video (right-click)
    //     <br/>
    //     <br/>
    // </div>
    var downloadImg = document.createElement("img");
    downloadImg.setAttribute("class", "srfchaddon");
    downloadImg.setAttribute("src", browser.runtime.getURL("images/rsich_64.png"));

    var downloadA   = document.createElement("a");
    downloadA.setAttribute("href", iframe.src);
    downloadA.dataset.urn = urn;

    // set text of link
    var downloadAtext = document.createTextNode(" " + browser.i18n.getMessage("bannerText"));
    downloadA.appendChild(downloadAtext);

    var downloadDiv = document.createElement("div");
    downloadDiv.setAttribute("style", "text-align:center");
    if (bannerId)
    {
        downloadDiv.setAttribute("id", bannerId);
    }
    downloadDiv.appendChild(downloadImg);
    downloadDiv.appendChild(downloadA);

    nodeParent.insertBefore(downloadDiv, divIframe);
}

/*
    addSrfPlayBanner():
    On some srf.ch pages (in particular srf.ch/play/...) the ID of the video is directly
    accessible in the address of the page. However, to let the add-on work also in this
    case a banner needs to be added such that the "normal" functions can be used
*/
function addSrfPlayBanner()
{
    // check if this event has already fired and there's already a download link
    // node added to the page.
    var downloadUrlFound = 0;
    if (document.getElementsByClassName("srfchaddon").length > 0)
    {
        downloadUrlFound = 1;
    }

    // current tab seems to be a srf.ch page
    // try to find out if it's really an SRF-play page...
    var isSrfPlay = 0;
    var metaElements = document.getElementsByTagName("meta");
    if (!downloadUrlFound && metaElements && metaElements.length > 0)
    {
        for (var i=0; i<metaElements.length; i++)
        {
            var siteName = 0;
            var content = "";
            for (var k=0; k<metaElements[i].attributes.length; k++)
            {
                if (metaElements[i].attributes[k].name === "property"
                    && metaElements[i].attributes[k].value === "og:site_name")
                {
                    siteName = 1;
                }
                else if (metaElements[i].attributes[k].name === "content")
                {
                    content = metaElements[i].attributes[k].value;
                }
            }

            if (siteName && content === "Play SRF")
            {
                isSrfPlay = 1;
            }
        }
    }

    if (isSrfPlay)
    {
        var nestedRowElement = document.getElementsByClassName("detailPageLeftCol");
        if (!(nestedRowElement && nestedRowElement.length==1))
        {
            nestedRowElement = document.getElementsByClassName("nestedRow");
        }

        if (nestedRowElement && (nestedRowElement.length > 0))
        {
            // add the following HTML code to the current document's DOM tree:
            // <div style="text-align:center">
            //     <img class="srfchaddon" src="http://addons.cdn.mozilla.net/user-media/addon_icons/413/413748-64.png" />
            //       Download URLs for this video (right-click)
            //     <br/>
            //     <br/>
            // </div>
            var downloadImg = document.createElement("img");
            downloadImg.setAttribute("class", "srfchaddon");
            downloadImg.setAttribute("src", browser.runtime.getURL("images/srfch_64.png"));

            var downloadA   = document.createElement("a");
            downloadA.setAttribute("href", document.documentURI);

            // set text of link
            var downloadAtext = document.createTextNode(" " + browser.i18n.getMessage("bannerText"));
            downloadA.appendChild(downloadAtext);

            var downloadDiv = document.createElement("div");
            downloadDiv.setAttribute("style", "text-align:center");
            downloadDiv.appendChild(downloadImg);
            downloadDiv.appendChild(downloadA);

            // (two) bracket returns
            downloadDiv.appendChild(document.createElement("br"));
            downloadDiv.appendChild(document.createElement("br"));

            var rowParent = nestedRowElement[0].parentElement;
            rowParent.insertBefore(downloadDiv, nestedRowElement[0]);
        }
    }
}